.gitlab-ci.yml

```diff
@@ -0,0 +1,100 @@
+##############################################################
+# Copyright 2023 Lawrence Livermore National Security, LLC
+# (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+#
+# This file is part of the Flux resource manager framework.
+# For details, see https://github.com/flux-framework.
+#
+# SPDX-License-Identifier: LGPL-3.0
+##############################################################
+
+stages: 
+  - test
+
+## YAML Anchors
+.build-core: &build-core
+    - git clone https://github.com/flux-framework/flux-core
+    - cd flux-core
+    # Some tests parse the most recent tag for testing.
+    - lstopo --of xml >$(hostname).xml
+    - export FLUX_HWLOC_XMLFILE=$(pwd)/$(hostname).xml
+    - ./autogen.sh
+    - ./configure
+    - make -j 32
+    - export PATH=$(pwd)/src/cmd:$PATH
+    - cd ..
+
+## Reusable Scripts
+.standard-variables:
+    variables:
+        LLNL_SERVICE_USER: fluxci
+        PYTHON: "/usr/bin/python3"
+        HWLOC_COMPONENTS: x86
+        debug: t
+        FLUX_TESTS_LOGFILE: t
+        FF_ENABLE_JOB_CLEANUP: "false"
+        CUSTOM_CI_BUILDS_DIR: "/usr/WS1/$$USER/gitlab-runner-builds-dir"
+        # Note: the above will not work with /usr/workspace, you must specify WS1 or WS2
+
+.test-core:
+    extends: .standard-variables
+    script:
+        - *build-core
+        - cd flux-core
+        - which flux
```

Comment:
At this point, the flux should be the just-built flux, because of the `export` in line 23. From a logfile on poodle:

```
[32;1m$ which flux[0;m
/usr/WS1/fluxci/gitlab-runner-builds-dir/fluxci/builds/5Y4Xcwgb/002/gitlab/hobbs17/flux-test-collective/flux-core/src/cmd/flux
```

Reply:
Ah, ok. 

Note that should not be necessary for `make check`, since all sharness tests effectively do the same at startup. It could also get confusing if you're testing a built flux on a Flux managed cluster, you don't want to use the built Flux to try to talk to the system instance since it could fail in strange ways... So I'm not convinced that prepending to the PATH here is doing anything useful, though I should ask why you thought it was necessary here?

Reply:
> So I'm not convinced that prepending to the PATH here is doing anything useful, though I should ask why you thought it was necessary here?

For our MPI testing, we discussed running this with a shell script instead of sharness. `.build-core` will be reused by multiple projects and testing scenarios, not all of which are guaranteed to use sharness, although this is certainly the case most of the time (and all of the time now).

We could fix the `PATH` in said shell script, but at that point. we're outside the `flux-core` directory. It probably wouldn't be hard to find the binaries but I went ahead and exported them after compilation.

> It could also get confusing if you're testing a built flux on a Flux managed cluster, you don't want to use the built Flux to try to talk to the system instance since it could fail in strange ways

The built flux _will_ fail if trying to allocate nodes because it's not the system instance owner but I've only hit that in my own manual testing. Did you have a specific use case in mind for talking to the system instance once the nodes are allocated?

If preferred, we could store the built flux path in `BUILT_FLUX`, and later prepend to the `PATH` if needed for MPI testing.

Reply:
> For our MPI testing, we discussed running this with a shell script instead of sharness.

I'm probably missing a piece of the puzzle here. For clusters with a system Flux instance, the gitlab-ci runner will have allocated nodes and started a Flux instance for the CI using `flux alloc` from the _system_ right?
If so, in order to launch a Flux instance running the _built_ flux-brokers, we'll have to use the system-installed `flux(1)` to submit a job to run the built `flux-broker`, e.g. something like:

```console
$ flux submit -N1 -n1 /path/to/flux-core.git/src/cmd/flux start test_script.sh
```

If we push the built `flux` to the front of `PATH`, this may not work since `flux` would point to our built `flux` and not the `flux` for the instance we're running.

If we leave `PATH` alone, then `src/cmd/flux start` will take care of prepending the correct path to the front of it, and things will "just work" in `test_script.sh`. 

> The built flux will fail if trying to allocate nodes because it's not the system instance owner but I've only hit that in my own manual testing. Did you have a specific use case in mind for talking to the system instance once the nodes are allocated?

The main use case is if we want to launch a multi-node test instance using the built flux from within a job allocated from the system instance.

> If preferred, we could store the built flux path in BUILT_FLUX, and later prepend to the PATH if needed for MPI testing.

I don't think that's necessary since whenever you launch a Flux instance using the built `flux(1)` it will prepend to `PATH` for you. If you don't launch a Flux instance using the built flux-core, then you may not be testing what you think you're testing...

Reply:
> I don't think that's necessary since whenever you launch a Flux instance using the built flux(1) it will prepend to PATH for you. If you don't launch a Flux instance using the built flux-core, then you may not be testing what you think you're testing...

I was thinking that the system isn't going to just know the full `/path/to/flux-core.git/src/cmd/flux` without saving this somehow. There are IDs and other things appended to the path created under GitLab. We can rely on the assumption that `./flux-core/src/cmd/flux` after the `*build-core` script will work, but I try to avoid relying on relative directory calls in the gitlab runners. 

Anyway, I removed the `PATH` stuff, the relative calls have worked up until now and I have no reason to believe that will change.

> If so, in order to launch a Flux instance running the built flux-brokers, we'll have to use the system-installed flux(1) to submit a job to run the built flux-broker, e.g. something like

Should we be doing `flux submit -N1 -n1 /path/to/src/cmd/flux submit make -j 32 check`? Or were you just noting this for future MPI tests?

Reply:
> I was thinking that the system isn't going to just know the full /path/to/flux-core.git/src/cmd/flux without saving this somehow.

Yes, that is a good thought. In the sharness setup, we set `FLUX_BUILD_DIR` and `FLUX_SOURCE_DIR`, perhaps setting both of those would do what you want? Then in scripts that need to launch test instances of flux, we can use either `$FLUX_BUILD_DIR/src/cmd/flux`.

> Should we be doing flux submit -N1 -n1 /path/to/src/cmd/flux submit make -j 32 check

No, I don't think we need to do that for `make check` since it already launches test instances for each sharness test via `test_under_flux()`. However, if we want to run other tests that launch test instances, we'll need to do that there as some kind of setup operation, e.g. either `flux run [OPTIONS] $FLUX_BUILD_DIR/src/cmd/flux start test.sh` or just `$FLUX_BUILD_DIR/src/cmd/flux start -s SIZE` if multi-node tests are not necessary.

Reply:
Some of this might require more thought when we're testing multiple framework projects together in this repo. The simplest thing to do would be to install all the projects into the same prefix, however, since it seems like gitlab likes to move things around, I'm not sure if this will work. We can look into that when we get there.

Reply:
@grondo Let me know if the most recent branch resolves this. I only set `FLUX_BUILD_DIR`, and it seems to be working with sharness.